evm/p2p/lightchain.py

```diff
@@ -1,69 +1,270 @@
+                best = peer
+        return best
 
-    async def get_peer(self):
-        raise NotImplementedError("TODO")
+    async def run(self):
+        while not self._should_stop.is_set():
+            try:
+                await self.maybe_connect_to_more_peers()
+            except:  # noqa: E722
+                # Most unexpected errors should be transient, so we log and restart from scratch.
+                self.logger.error("Unexpected error (%s), restarting", traceback.format_exc())
+                await self.stop_all_peers()
+            # Wait self._connect_loop_sleep seconds, unless we're asked to stop.
+            await asyncio.wait([self._should_stop.wait()], timeout=self._connect_loop_sleep)
+        self._finished.set()
+
+    async def stop_all_peers(self):
+        self.logger.info("Stopping all peers ...")
+        await asyncio.gather(
+            *[peer.stop_and_wait_until_finished() for peer in self.connected_nodes.values()])
 
     async def stop(self):
-        raise NotImplementedError("TODO")
+        self._should_stop.set()
+        await self.stop_all_peers()
+        await self._finished.wait()
+
+    async def connect(self, remote):
+        """
+        Connect to the given remote and return a Peer instance when successful.
+
+        Returns None if the remote is unreachable, times out or is useless.
+        """
+        if remote in self.connected_nodes:
+            self.logger.debug("Skipping %s; already connected to it", remote)
+            return None
+        try:
+            peer = await asyncio.wait_for(
+                handshake(remote, self.privkey, self.peer_class, self.chaindb, self.network_id),
+                HANDSHAKE_TIMEOUT)
+            return peer
+        except (UnreachablePeer, asyncio.TimeoutError, PeerConnectionLost) as e:
+            self.logger.debug("Could not complete handshake with %s: %s", remote, e)
+        except UselessPeer:
+            self.logger.debug("No matching capabilities with %s", remote)
+        except PeerDisconnected as e:
+            self.logger.debug(
+                "%s disconnected before completing handshake; reason: %s", remote, e)
+        except Exception:
+            self.logger.warn(
+                "Unexpected error during auth/p2p handhsake with %s: %s",
+                remote, traceback.format_exc())
+        return None
+
+    async def maybe_connect_to_more_peers(self):
+        """Connect to more peers if we're not yet connected to at least self.min_peers."""
+        if len(self.connected_nodes) >= self.min_peers:
+            self.logger.debug(
+                "Already connected to %s peers: %s; sleeping",
+                len(self.connected_nodes),
+                [remote for remote in self.connected_nodes])
+            return
+
+        for node in await self.get_nodes_to_connect():
+            self.logger.debug("Connecting to %s", node)
+            peer = await self.connect(node)
+            if peer is not None:
```

Comment:
This is a pattern I really don't like (checking return values).  I suspect it may be this way due to `self.connect(...)` being something that get's `await`-ed.  Is it possible for this to instead use a try/except pattern, or some other pattern that removes the return value checking?

Reply:
When connecting to a peer there are various possible exceptions that can be raised, and I decided to handle them all in connect() to keep things simple here. I could change connect() to raise something else, though? Any name suggestions for such a new exception?

Reply:
I'd suggest establishing a common base exception, something like `PeerConnectionError` that all of the exceptions inherit from.  For the one `asyncio.TimeoutError` we can catch and re-raise our own exception (maybe `PeerTimeoutError`).

Reply:
One thing to keep in mind is that we'll want to handle each of those exceptions differently; for example, a UselessPeer should probably not be retried any time soon, whereas a TimeoutError should. That's why I wanted to keep the handling of them in `connect()`. We could have `connect()` raise a `PeerConnectionError`, but that wouldn't feel right in the case of a `UselessPeer` (i.e. a peer with no matching sub-protocols)

Reply:
I'm comfortable with just adding a `# TODO` comment here to explore ways to remove return-value-checking from this API.